Lib/datetime.py

```diff
@@ -262,86 +262,185 @@ def _wrap_strftime(object, format, timetuple):
     return _time.strftime(newformat, timetuple)
 
 # Helpers for parsing the result of isoformat()
+def _find_isoformat_separator(dtstr):
+    # See the comment in _datetimemodule.c:_findisoformat_separator
+    len_dtstr = len(dtstr)
+    if len_dtstr == 7:
+        return 7
+
+    assert len_dtstr > 7
+    date_separator = "-"
+    week_indicator = "W"
+
+    if dtstr[4] == date_separator:
+        if dtstr[5] == week_indicator:
+            if len_dtstr < 8:
+                raise ValueError("Invalid ISO string")
+            if len_dtstr > 8 and dtstr[8] == date_separator:
+                if len_dtstr == 9:
+                    raise ValueError("Invalid ISO string")
+                if len_dtstr > 10 and dtstr[10].isdigit():
+                    # This is as far as we need to resolve the ambiguity for
+                    # the moment - if we have YYYY-Www-##, the separator is
+                    # either a hyphen at 8 or a number at 10.
+                    #
+                    # We'll assume it's a hyphen at 8 because it's way more
+                    # likely that someone will use a hyphen as a separator than
+                    # a number, but at this point it's really best effort
+                    # because this is an extension of the spec anyway.
+                    # TODO(pganssle): Document this
```

Comment:
This is a constraint we can't really change without a deprecation warning, because it's left over from the "`datetime.fromisoformat` parses everything that `datetime.isoformat` outputs" era. `datetime.isoformat(sep="5")` is a valid but extremely weird thing to do, so we have to be able to parse it.

I suppose we could limit the ability to use a number as the separator to cases that could actually come from `.isoformat()`, though.